### ISSUE TITLE
design(import): the inbound half, audio and MIDI to MuseCode material

### ISSUE THREAD

**Issue by Soushi888** [Soushi888/music-as-code#52]
> **SoushAI analysis.** Drafted by Soushi's AI assistant, reviewed and posted by @Soushi888.

## Context

Everything in the repo today points outward. `resolve` projects a `Music` tree to events, `backends::midi` writes, `analysis` reads events this project itself produced, and the M2/M3 backends add more ways out. Nothing reads in. Soushi wants the other direction: take an audio file or a MIDI file of a piece he likes, pull out its ideas, and have them arrive as MuseCode material he can compose with.

The research pass of 2026-09-14 (four parallel surveys: automatic music transcription, classic MIR feature extraction, symbolic structure analysis, and the Suno/DAW round trip) settled what is actually achievable. The headline is an asymmetry that should shape the whole design.

| Task, from a real produced recording | Best available 2026 result | Source |
|---|---|---|
| Beat and downbeat | F1 97.5 / 95.3 (Ballroom), 62.7 on the hard expressive set | Beat This!, ISMIR 2024 |
| Section boundaries | HR.5F 0.660 (Harmonix) | All-In-One, WASPAA 2023 |
| Key, 24 classes only | roughly 75 to 84%, cannot emit a modal label | Temperley; madmom key CNN |
| Chords, 25 classes only | major/minor/no-chord; no 7b9, no slash, no quartal | madmom, autochord, Chordino |
| Multi-instrument notes | **Multi F1 47.8** | MuScriptor, arXiv:2607.08168, July 2026 |
| Solo piano notes, studio audio | note F1 97.4 | hFT-Transformer, ISMIR 2023 |
| One monophonic line (bass, lead) | 78.8 F on real jazz bass | CREPE Notes, ISMIR 2023 |

Two hard ceilings worth stating once. Every chord recogniser verified in the survey is capped at twelve major plus twelve minor plus no-chord, so no MIR tool will hand back the harmony this project actually cares about. Every key detector is a 24-class softmax, so modal material gets force-fit to the nearest major or minor. The interesting harmony has to come from this repo's own `theory` layer reading resolved events, not from a model.

## The reframe

This is not a transcriber and should never try to be. At 47.8 Multi F1 on a full mix, an import is a sketch. That is fine, and Soushi confirmed it on 2026-09-15: the point is to **extract ideas from an input and organise them by hand into a MuseCode piece**, not to reconstruct a score. Audio and MIDI are two inputs to the same inspiration pipeline, MIDI being the exact one and audio the fuzzy one.

That reframe also says where the value is. The scaffold (tempo map, downbeats, sections, groove) is 85 to 95% reliable. The notes are not. So the import path should be confident about the scaffold, honest about the notes, and always leave the composing to the human. This is the same instinct already written into #45: the shelf grows by ear.

## Architecture: the lift is `resolve` inverted

`resolve()` maps `Music -> Resolved { events, tempo_map, time_sigs }`, and `Event` already carries onset, dur, written_dur, pitch, velocity, voice, instrument, articulation. That is exactly the shape a transcriber or a MIDI parser emits. So the inbound problem is the inverse of a pass this repo already owns: a **lift** from `Vec<Event>` to `Music`.

Projection is deterministic. Lifting is an inverse problem with many valid answers, and the job of MIR is to supply the context (beat grid, tempo curve, key, sections) that picks one.

That gives the correctness law, the sibling of the parser round-trip law in #14:

```
resolve(lift(resolve(m))) == resolve(m)
```

Not `lift(resolve(m)) == m`, which is false and should be: the lift cannot know that `F4` was written `d!(1)`. But the resolved projection must be stable under a round trip, and that is a property test over generated trees. It is the single best acceptance criterion for the whole inbound side, and it plays the role the resolver played in M1.

## Decisions settled 2026-09-15

**1. The lift emits `Chromatic`. Degrees are an explicit second pass.** `lift` never guesses a key into the tree. A separate `.to_degrees(key)` converts chromatic pitches to `Degree` against a key the caller names or that key-finding proposed. The guess stays visible and refusable, and an import that is wrong about the key is still correct material.

**2. The Python sidecar is external, input and output both.** Nothing links Python into the Rust binary, and there is no live protocol. The sidecar is a separate program: audio file in, a JSON analysis document out. `musecode_import` reads that document as a file, exactly as it reads a `.mid`. This drops the localhost HTTP idea from the brainstorm in favour of the simpler contract, and it means the Rust crates build and test with no Python present.

**3. The lift is lossy on purpose.** No completeness goal, no fidelity metric on dense material. An import that returns a correct tempo map, correct downbeats, plausible sections, one good bass line and three repeated motifs has succeeded, even if it missed most of the inner voices.

## The ladder

Six rungs, each a strictly harder inverse, each independently shippable, each useful alone.

| Rung | From, to | How | Rust today |
|---|---|---|---|
| 0 | MIDI file, `Resolved` | `midly` reads as well as writes; `delta_ticks / PPQN` is exactly a `Rational32`, so tick conversion is exact with no float; tempo meta-events become the tempo map; type-1 tracks become `Par` of `Seq`, one voice each | already a dependency; no crate does the musical bookkeeping, roughly 300 lines |
| 1 | performance events, a grid | Cemgil's switching state-space model: discrete note-location switches over a continuous hidden tempo, tempo tracking as filtering and quantization as MAP decoding by particle filter (arXiv:1106.4863) | nothing exists, implement from the paper |
| 2 | events, `Music` with degrees | Krumhansl-Schmuckler key finding is a correlation against twelve rotations of a profile vector: deterministic, explainable, no dependency. Feeds the explicit `.to_degrees(key)` pass | trivial port |
| 3 | events, `Pattern` | No MIR tool anywhere emits a readable rhythmic signature, which is a genuine gap in the field. This repo does not need one: it already has `Pattern`, `Step`, `tresillo`, `habanera`, `cinquillo`, `euclid` and `from_grid`, so matching onsets against its own grid vocabulary is string distance over `x..x..x...x.x...` | own build, on existing types |
| 4 | events, a shelf | COSIATEC (Meredith): music as a point set in (onset, pitch), find all Maximal Translatable Patterns, group into Translational Equivalence Classes, greedily take the most compressive and repeat. The output *is* a fragment shelf, and one TEC maps onto one `Phrase` with its occurrences as transposition offsets | own build; Blake3 content addressing already fits |
| 5 | a corpus, a style | Conklin's multiple viewpoint systems (JNMR 1995): a style is an ensemble of small per-viewpoint distributions (pitch, interval, duration, metrical position) with a piece-local and a corpus-wide model. Distributions, not a checkpoint: serialisable and inspectable | own build; feeds the `Style` component of #44 |

## Groove is the rung that pays for the others

`swing` (#20) and `humanize` (#21) are declared and both `todo!()`. Groove, in the Magenta/GrooVAE sense, is nothing more than a per-grid-position table of velocity and timing deviation from the nearest grid point. Extract that table from a recording and it becomes a named, inspectable data value:

```rust
let feel = groove::from_midi("libertango.mid")?;   // a table, not a model
melody.pipe(humanize(feel))
```

That turns import from a novelty into the thing that unblocks two open issues in M3, and it is the honest answer to "capture the style": the style that can be captured reliably is the groove and the form, not the notes.

## Crate boundary

None of this belongs in `musecode_core`; the layer rules in CLAUDE.md would not survive it. Two new crates, following the precedent set by #49:

- **`musecode_import`**: pure Rust, no ML, no network. Rungs 0 to 5, plus reading the sidecar's JSON analysis document. Depends on `musecode_core`; nothing in core depends on it.
- **`musecode_listen`**: audio. `symphonia` to decode, `beat-this` for the scaffold, basic-pitch through `ort` for notes. Depends on `musecode_import`.

## Audio: what actually runs natively in Rust

Better than expected. The scaffold, which is the reliable half, runs entirely in Rust.

| Layer | Rust status |
|---|---|
| Decode | `symphonia`, pure Rust, MP3/FLAC/WAV/AAC |
| FFT | `rustfft` / `realfft`, production grade |
| Beat and downbeat | **`beat-this` crate, MIT, native port of the 2024 state of the art, on `rten` or ONNX** |
| Note transcription | no AMT crate, but basic-pitch's ONNX runs in Rust today through `ort` (precedent: `ja-ortiz-uniandes/open-piano`) |
| Onsets | `microdsp`, MIT, native spectral flux |
| Chords, structure, timbre embeddings | nothing native; the sidecar's job |

basic-pitch is the right first audio dependency: Apache-2.0, under 17K parameters, ships ONNX in its own repo, and it is the only mainstream system that emits pitch bend as a first-class head. Everything in the MT3 lineage, MuScriptor included, token-quantises onsets to roughly 10ms bins, which discards exactly the microtiming this project cares about.

## Acceptance criteria (epic)

- [ ] An ADR in `docs/design/` settles the `lift` contract, the crate split, and the sidecar document schema before any rung ships
- [ ] `resolve(lift(resolve(m))) == resolve(m)` holds as a property test over generated trees
- [ ] A `.mid` file imports to a `Music` tree that prints readable notation through the existing `Display` and plays back through `just play`
- [ ] One audio file Soushi chooses produces a correct tempo map, correct downbeats, and at least one usable fragment on the shelf, verified by ear
- [ ] Both new crates build and test with no Python installed
- [ ] `swing` and `humanize` consume an extracted groove table rather than a hardcoded ratio

## Out of scope

Score-accurate transcription of dense mixes. Any fidelity metric on a full-band import. Live or real-time input. Lyrics or vocals. Training any model. Shipping model weights in the repo (note that MuScriptor's weights are CC BY-NC 4.0 and madmom's pretrained models are CC BY-NC-SA 4.0, so neither can be redistributed here; basic-pitch is Apache-2.0 and `beat-this` is MIT).

## The round trip, for the record

Out: MuseCode to `.mid`, then fluidsynth today or oxisynth (#24), or Reaper's `-batchconvert` for a scripted render. Into Suno Studio, which splits a track into up to twelve time-aligned WAV stems and extracts a `.mid` from a stem on the Premier tier. Back in: NeuralNote (free, open, VST3/CLAP/standalone, Linux, built on the same basic-pitch model) is the audio-to-MIDI tool of record on this platform. Samplab is sunsetting 2026-09-17 and should not be built on. Udio disabled downloads after its label settlement and is not usable in a round trip.

One unresolved point: whether Suno Studio still accepts a MIDI **upload** as a generation input. A fetch of help.suno.com/en/articles/13670593 on 2026-09-14 described only creating MIDI inside Studio, with no mention of import. If MIDI upload is gone, the outbound bridge is audio-only.

## Legal note

Analysing a recording and keeping a private transcription sits inside Canadian fair dealing for research and private study (Copyright Act s.29). Publishing a transcription is a fresh argument on the fairness factors; publishing derived audio is the riskiest step; training on copyrighted audio is unresolved policy with no Canadian legislation yet. Style and genre are ideas and not copyrightable; a fixed melody is expression. Informational, not legal advice. Practical consequence for this repo: imported material from copyrighted recordings stays out of committed examples and out of the shipped fragment shelf.

## Related

- Unblocks: #20 (swing), #21 (humanize), both of which want a groove table rather than a constant
- Feeds: #44 (components carrying style), #45 (the shelf by ear), #46 (the fragment catalog)
- Uses: #14 (`.muse` text syntax is the natural output format for an import), #8 (`Display` gives a readable transcription for free)
- Precedent: #49 (splitting a concern into its own crate)
